### PR TITLE
feat(filetype.lua): add support for tmux.conf files

### PR DESCRIPTION
**Pull request by gsx64**
I found that tmux.conf was not detected after opt-in filetype.lua.
I made some changes to filetype.lua after referring [filetype.vim](https://github.com/neovim/neovim/blob/master/runtime/filetype.vim#L1887).

### CODE REVIEW

**Comment by zeertzjq** [comment]
The commit message needs to follow <https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#commit-messages>

File: runtime/lua/vim/filetype.lua [hunk 1]
```diff
@@ -1226,6 +1227,7 @@ local pattern = {
   [".*/%.config/systemd/user/.*%.d/.*%.conf"] = "systemd",
   [".*/etc/systemd/system/.*%.d/.*%.conf"] = "systemd",
   [".*%.t%.html"] = "tilde",
+  [".?tmux.*%.conf.*"] = "tmux",
```
Comment (gpanders) [inline runtime/lua/vim/filetype.lua:1230]: The leading `.` needs to be escaped. And the pattern in `filetype.vim` doesn't have a `*` at the end.

```suggestion
  ["%.?tmux.*%.conf"] = "tmux",
```

Comment (ichigozero) [reply runtime/lua/vim/filetype.lua:1230]: Thank you for the review.

I doubled checked `filetype.vim` and I found that there are two entries for tmux.

- [line 1887](https://github.com/neovim/neovim/blob/master/runtime/filetype.vim#L1887)
  ```vim
  au BufNewFile,BufRead {.,}tmux*.conf		setf tmux
  ```

- [line 2367](https://github.com/neovim/neovim/blob/master/runtime/filetype.vim#L2367)
  ```vim
  au BufNewFile,BufRead {.,}tmux*.conf*		setf tmux
  ```

In line 2367, the pattern does have `*` at the end. I wonder which one is correct.

Comment (gpanders) [reply runtime/lua/vim/filetype.lua:1230]: Good catch, it looks like this is indeed intentional: https://github.com/vim/vim/commit/e519eb41c1c12836b2d12aeb703bb04c7618a724

Comment (zeertzjq) [reply runtime/lua/vim/filetype.lua:1230]: <https://github.com/vim/vim/pull/8971#issuecomment-939315001>

> Adding a star at the end of a filetype pattern is tricky, because it
> matches other things, such as ".gz" or ".old".  It should always be
>checked last.

Comment (zeertzjq) [reply runtime/lua/vim/filetype.lua:1230]: This can be complicated. Can `filetype.lua` currently achieve this?

Comment (ichigozero) [reply runtime/lua/vim/filetype.lua:1230]: I see. If that is the case should I make two entries such as the following

```lua
["%.?tmux.*%.conf"] = "tmux",
["%.?tmux.*%.conf.*"] = "tmux",
```

where the second entry is put further down to avoid cases such as `.tmux.conf.gz` is recognized as `tmux` filetype instead of `gzip` filetype.

Or is one entry `["%.?tmux.*%.conf.*"] = "tmux"` enough?

Comment (zeertzjq) [reply runtime/lua/vim/filetype.lua:1230]: I think Lua tables are unordered?

Comment (ichigozero) [reply runtime/lua/vim/filetype.lua:1230]: I just started learning Lua recently so I didn't know about Lua tables being unordered.

To keep in line `filetype.vim`, I think one entry of `["%.?tmux.*%.conf.*"] = "tmux"` is more appropriate than `["%.?tmux.*%.conf"] = "tmux"`.

Is my understanding correct?

Comment (zeertzjq) [reply runtime/lua/vim/filetype.lua:1230]: I'm not sure about this either.

Comment (gpanders) [reply runtime/lua/vim/filetype.lua:1230]: Hmm, I wonder why Bram didn't want to use `s:StarSetf` then? (e have our own `starsetf` function to match those). From `filetype.vim`:

```
" Extra checks for when no filetype has been detected now.  Mostly used for
" patterns that end in "*".  E.g., "zsh*" matches "zsh.vim", but that's a Vim
" script file.
" Most of these should call s:StarSetf() to avoid names ending in .gz and the
" like are used.
```

In any case, we can enforce priority in the patterns with the priority key. So what we want is

```lua
["%.?tmux.*%.conf"] = "tmux",
["%.?tmux.*%.conf.*"] = { "tmux", { priority = -1 } }
```

Comment (ichigozero) [reply runtime/lua/vim/filetype.lua:1230]: Thank you for the clarification. I will push the new changes.

**Review by gpanders — APPROVED** [review]

File: runtime/lua/vim/filetype.lua (hunk 1 above)
Comment (clason) [reply runtime/lua/vim/filetype.lua]: Lua tables are indeed unordered; one of the goals of this refactor is to move away from relying on (hard to follow) ordering and instead using explicit priorities, as in
```lua
[".*/etc/foo/.*%.conf"] = { "dosini", { priority = 10 } },
```

Comment (ichigozero) [reply runtime/lua/vim/filetype.lua]: I see. Thank you for the explanation.

**Comment by gpanders** [comment]
Included in #16978.
